Archive/Wiedijk100Theorems/BuffonsNeedle.lean

```diff
@@ -0,0 +1,450 @@
+
+/--
+  Projection of a needle onto the x-axis. The needle's center is at
+  x-coordinate `x`, of length `l` and angle `θ`. Note, `θ` is measured
+  relative to the y-axis, that is, a vertical needle has `θ = 0`.
+-/
+def needle_x_proj (x θ : ℝ) : Set ℝ := Set.Icc (x - θ.sin * l / 2) (x + θ.sin * l / 2)
+
+/--
+  A random variable representing whether the needle crosses a line.
+
+  The line is at `x = 0`, and the needle crosses the line if it's projection
+  onto the x-axis contains `0`. This random variable is `1` if the needle
+  crosses the line, and `0` otherwise.
+
+  Note: `N : Ω → ℝ` is the random variable; the definition of `N' : ℝ × ℝ` is
+  provided for convenience.
+-/
+noncomputable def N' (p : ℝ × ℝ) : ℝ := Set.indicator (needle_x_proj l p.1 p.2) 1 0
+noncomputable def N : Ω → ℝ := N' l ∘ B
+
+lemma short_needle_inter_eq (h : l ≤ d) (θ : ℝ) :
+    Set.Icc (-d / 2) (d / 2) ∩ Set.Icc (-θ.sin * l / 2) (θ.sin * l / 2) =
+    Set.Icc (-θ.sin * l / 2) (θ.sin * l / 2) := by
+
+  rw [Set.Icc_inter_Icc, inf_eq_min, sup_eq_max, max_div_div_right zero_le_two,
+    min_div_div_right zero_le_two, neg_mul, max_neg_neg, mul_comm,
+    min_eq_right (mul_le_of_le_of_le_one_of_nonneg h θ.sin_le_one hl.le)]
+
+abbrev B_range := Set.Icc (-d / 2) (d / 2) ×ˢ Set.Icc 0 π
+
+lemma B_range_volume : ℙ (B_range d) = ENNReal.ofReal (d * π) := by
+  simp_rw [MeasureTheory.Measure.volume_eq_prod, MeasureTheory.Measure.prod_prod, Real.volume_Icc]
+  ring_nf
+  exact (ENNReal.ofReal_mul hd.le).symm
+
+lemma B_range_nonzero : ℙ (B_range d) ≠ 0 := by
+  rw [B_range_volume d hd, ne_eq, ENNReal.ofReal_eq_zero, not_le]
+  exact mul_pos hd Real.pi_pos
+
+lemma B_range_nontop : ℙ (B_range d) ≠ ⊤ := by
+  rw [B_range_volume d hd]
+  exact ENNReal.ofReal_ne_top
+
+lemma B_range_measurable : MeasurableSet (B_range d) :=
+  MeasurableSet.prod measurableSet_Icc measurableSet_Icc
+
+instance instBHasPDF : MeasureTheory.HasPDF B ℙ :=
+  MeasureTheory.pdf.IsUniform.hasPDF
+    (B_range_nonzero d hd) (B_range_nontop d hd) hB
+
+lemma N'_measurable : Measurable (N' l) := by
+  unfold N'
+
+  apply Measurable.indicator measurable_const
+  apply IsClosed.measurableSet
+  apply IsClosed.inter
+
+  case' h.h₁ =>
+    simp only [tsub_le_iff_right, zero_add]
+    apply isClosed_le continuous_fst _
+  case' h.h₂ =>
+    simp only [← neg_le_iff_add_nonneg']
+    apply isClosed_le _ _
+
+  · conv => arg 1; intro p; rw [← mul_one p.1, ← mul_neg]
+    exact Continuous.mul continuous_fst continuous_const
+
+  all_goals
+  · simp_rw [mul_div_assoc]
+    apply Continuous.mul _ continuous_const
+    have : (fun (x : ℝ × ℝ) => Real.sin x.2) = Real.sin ∘ Prod.snd := by rfl
+    rw [this]
+    apply Continuous.comp Real.continuous_sin continuous_snd
+
+lemma N'_strongly_measurable : MeasureTheory.StronglyMeasurable (N' l) := by
+  apply stronglyMeasurable_iff_measurable_separable.mpr
+  apply And.intro
+  · exact N'_measurable l
+  · apply Exists.intro {0, 1}
```

Comment:
This is the sort of thing that `refine'` is really good for. Does something like this work? (untested)
`refine' stronglyMeasurable_iff_measurable_separable.mpr ⟨N'_measurable l, {0, 1}, _⟩`